server/src/server.ts

```diff
@@ -0,0 +1,344 @@
+		capabilities.textDocument.publishDiagnostics.relatedInformation
+	);
+	const result: InitializeResult = {
+		capabilities: {
+			textDocumentSync: TextDocumentSyncKind.Incremental,
+			documentSymbolProvider: true,
+			definitionProvider: true,
+		}
+	};
+	if (hasWorkspaceFolderCapability) {
+		result.capabilities.workspace = {
+			workspaceFolders: {
+				supported: true
+			}
+		};
+	}
+	// console.log('LPython language server initialized');
+	return result;
+});
+
+connection.onInitialized(() => {
+	if (hasConfigurationCapability) {
+		// Register for all configuration changes.
+		connection.client.register(DidChangeConfigurationNotification.type, undefined);
+	}
+	if (hasWorkspaceFolderCapability) {
+		// eslint-disable-next-line @typescript-eslint/no-unused-vars
+		connection.workspace.onDidChangeWorkspaceFolders(_event => {
+			// connection.console.log('Workspace folder change event received.');
+		});
+	}
+});
+
+// Borrowed from Jakt: https://github.com/SerenityOS/jakt
+function convertPosition(position: Position, text: string): number {
+	let line = 0;
+	let character = 0;
+	const buffer = new TextEncoder().encode(text);
+
+	let i = 0;
+	while (i < text.length) {
+	  if (line == position.line && character == position.character) {
+		return i;
+	  }
+
+	  if (buffer.at(i) == 0x0A) {
+		line++;
+		character = 0;
+	  } else {
+		character++;
+	  }
+
+	  i++;
+	}
+
+	return i;
+  }
+
+  connection.onDefinition(async (request) => {
+	console.time('onDefinition');
+	const document = documents.get(request.textDocument.uri);
+	const settings = await getDocumentSettings(request.textDocument.uri);
+	const text = document?.getText();
+	const symbols: SymbolInformation[] = [];
+	if (typeof text == "string") {
+	  console.log("request: ");
+	  console.log(request);
+	  console.log("index: " + convertPosition(request.position, text));
+	  const stdout = await runCompiler(text, "--goto-def " + convertPosition(request.position, text), settings);
+	  console.log("got: ", stdout);
+	  const obj = JSON.parse(stdout);
+	  for (let i = 0; i < obj.length; i++) {
+		if (obj[i].location) {
+		  return [{
+			targetUri: request.textDocument.uri,
+			targetRange: {
+			  start: { line: obj[i].location.range.start.line, character: obj[i].location.range.start.character },
+			  end: { line: obj[i].location.range.end.line, character: obj[i].location.range.end.character }
+			},
+			targetSelectionRange: {
+			  start: { line: obj[i].location.range.start.line, character: obj[i].location.range.start.character },
+			  end: { line: obj[i].location.range.end.line, character: obj[i].location.range.end.character }
+			},
+			originSelectionRange: {
+			  start: { line: request.position.line, character: Math.max(0, request.position.character - 4) },
+			  end: { line: request.position.line, character: request.position.character + 4 }
+			}
+		  }];
+		}
+	  }
+	  // console.timeEnd('onDefinition');
+	  return undefined;
+	}
+  });
```

Comment:
This is the relevant code for this PR. Once previous PRs are merged, and this is rebased, the PR will look cleaner.